### PR TITLE
ssr middleware

### PR DESCRIPTION
**Pull request by dratwas**
This PR contains SSR middleware. Its not done but its working so shouldn't be a blocker.

TODO: 
- [x] render html, body, entry 
- [x] render head with assets
- [x] caching
- [x] error handling
- [x] ReactHotAPI is not a function
- [x] handler rebuilding server 2 times
- [x] render email

TODO ALSO:
- [x] refactor errorHandler

For working with it we need file with entries in project. I suggest to add `entries.js` to `src/config/` directory.

Also here i want to discuss about dir structure. I think we could create directory called entries in src directory. Something like that

Hooks, flow and test will be in separate PR.

```
src
│   Index.js
└───config
│   │   entries.js
│   │   ...
└───entries
    │   home.js
    │   ...
```

### CODE REVIEW

**Comment by zamotany** [comment]
Moved structure disscussion to issue #560

File: packages/gluestick/src/config/compileWebpackConfig.js [hunk 1]
```diff
@@ -19,7 +19,9 @@ module.exports = (
   plugins: Plugin[],
   projectConfig: ProjectConfig,
   gluestickConfig: GSConfig,
-  { skipEntryGeneration }: { skipEntryGeneration: boolean } = {},
+  { skipClientEntryGeneration, skipServerEntryGeneration }: {
```
Comment (michalchudziak) [inline packages/gluestick/src/config/compileWebpackConfig.js]: Maybe extract type for better readability?

File: packages/gluestick/src/config/webpack/webpack.config.client.dev.js [hunk 2]
```diff
@@ -32,9 +32,16 @@ module.exports = (
       ],
     });
   }, {});
-  configuration.module.rules[0].use.unshift({
-    loader: 'react-hot-loader',
-  });
+  configuration.module.rules[0].use[0].options.plugins.push([
```
Comment (michalchudziak) [inline packages/gluestick/src/config/webpack/webpack.config.client.dev.js]: Looks a bit hacky, maybe we could extract something or add comment.

File: packages/gluestick/src/renderer/helpers/cacheManager.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+const LRU = require('lru-cache');
+
+// Creatin cache
+const DEFAULT_TTL: number = 60 * 60 * 1000;
+const lruOptions: { maxAge: number, max: number } = {
+  maxAge: DEFAULT_TTL,
+  max: 50,
+};
+const _cache: Object = LRU(lruOptions);
```
Comment (michalchudziak) [inline packages/gluestick/src/renderer/helpers/cacheManager.js:13]: Why underscore 😇?

File: packages/gluestick/src/renderer/helpers/errorHandler.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+// We use handlebars to deliver the 505 page. This lets you
+// use a handlebars template so you can display the stack trace or
+// any request, response information you would like.
+Handlebars.registerHelper('notForProduction', (options) => {
+  return process.env.NODE_ENV === 'production' ? '' : options.fn(this);
+});
+
+const tryReadFile = (filePath: string): Promise<string | boolean> =>
+  new Promise((resolve, reject) => {
+    fs.readFile(filePath, 'utf8', (error, data) => {
```
Comment (michalchudziak) [inline packages/gluestick/src/renderer/helpers/errorHandler.js]: Missing types.

File: packages/gluestick/src/renderer/helpers/errorHandler.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+  res: Response,
+  error: { status?: number },
+  custom505FilePath: string,
+  gluestick505FilePath: string,
+): Promise<void> => {
+  res.status(error.status || 500);
+  try {
+    const customTemplate = await tryReadFile(custom505FilePath);
+    const output = customTemplate || await tryReadFile(gluestick505FilePath);
+    if (output) res.send(Handlebars.compile(output)({ req, res, error }));
```
Comment (michalchudziak) [inline packages/gluestick/src/renderer/helpers/errorHandler.js]: Add bracers :>

File: packages/gluestick/src/renderer/helpers/getAssetsLoader.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
+/* @flow */
+
+const serialize = require('serialize-javascript');
+
+module.exports = (loadjsConfig: Object, entryPointBundle: string, vendorBundle: string): string => `
```
Comment (michalchudziak) [inline packages/gluestick/src/renderer/helpers/getAssetsLoader.js]: Maybe add some comment :)?

**Review by michalchudziak — DISMISSED** [review]

**Comment by michalchudziak** [comment]
Generally we are missing typings in quite some files in this PR. Can you sort it out somehow 😇?

File: packages/gluestick/src/renderer/helpers/errorHandler.js (hunk 5 above)
Comment (zamotany) [reply packages/gluestick/src/renderer/helpers/errorHandler.js]: Taken care of in #591

**Comment by zamotany** [comment]
@mike866  https://github.com/TrueCar/gluestick/pull/591

File: packages/gluestick/src/renderer/helpers/cacheManager.js (hunk 3 above)
Comment (zamotany) [reply packages/gluestick/src/renderer/helpers/cacheManager.js:13]: Why not? :P

File: packages/gluestick/src/renderer/helpers/getAssetsLoader.js (hunk 6 above)
Comment (zamotany) [reply packages/gluestick/src/renderer/helpers/getAssetsLoader.js]: Taken care of in #591

File: packages/gluestick/src/renderer/helpers/errorHandler.js (hunk 4 above)
Comment (zamotany) [reply packages/gluestick/src/renderer/helpers/errorHandler.js]: Taken care of in #591

File: packages/gluestick/src/config/webpack/webpack.config.client.dev.js (hunk 2 above)
Comment (zamotany) [reply packages/gluestick/src/config/webpack/webpack.config.client.dev.js]: Taken care of in #591

File: packages/gluestick/src/config/compileWebpackConfig.js (hunk 1 above)
Comment (zamotany) [reply packages/gluestick/src/config/compileWebpackConfig.js]: Taken care of in #591

File: packages/gluestick/src/renderer/helpers/cacheManager.js (hunk 3 above)
Comment (dratwas) [reply packages/gluestick/src/renderer/helpers/cacheManager.js:13]: @mike866 What is your suggestion?

**Review by michalchudziak — APPROVED** [review]

**Pull request merged**